src/wallet/spend.cpp

```diff
@@ -557,14 +559,28 @@ std::optional<SelectionResult> ChooseSelectionResult(const CWallet& wallet, cons
         results.push_back(*srd_result);
     }
 
-    if (results.size() == 0) {
+    if (results.empty()) {
         // No solution found
         return std::nullopt;
     }
 
+    std::vector<SelectionResult> eligible_results;
+    std::copy_if(results.begin(), results.end(), std::back_inserter(eligible_results), [coin_selection_params](const SelectionResult& result) {
+        const auto initWeight{coin_selection_params.tx_noinputs_size * WITNESS_SCALE_FACTOR};
+        const int weight = std::accumulate(result.GetInputSet().begin(), result.GetInputSet().end(), static_cast<int>(initWeight), [](int acc, const auto& coin) {
+            return acc + coin.input_bytes * WITNESS_SCALE_FACTOR;
+        });
```

Comment:
Instead of doing this entire for-loop accumulation, what about adding a field `m_weight_sum` to `OutputGroup`?

Initializing it to `coin_selection_params.tx_noinputs_size * WITNESS_SCALE_FACTOR` in the constructor and increasing it at every output insertion (inside `OutputGroup::Insert`).

So, when `SelectionResult::AddInput` gets called, you already have the total weight for each output group. And can just sum it up inside another cache field there.